### PR TITLE
dstor: avoid the use of macro function

### PR DESCRIPTION
**Pull request by karasevb**
Signed-off-by: Boris Karasev <karasev.b@gmail.com>

### CODE REVIEW

**Comment by karasevb** [comment]
refs https://github.com/open-mpi/ompi/pull/4606

**Comment by rhc54** [comment]
Is there some reason as to why all this code is declared inline? I'm asking because it tends to cause optimization issues, especially when the inlined sections are long and/or involve conditional jump (e.g., "if" and "switch") statements.

**Comment by karasevb** [comment]
I use inline, because those functions are called frequently, and some of them has inbuilt in other functions calls which in its turn can be invoked separately as well.
For example here: https://github.com/pmix/pmix/pull/609/files#diff-2dfe3f5a8c9f7e18a7a720a85ef91714R69
I guess it shouldn't be lead to degradation.

**Comment by rhc54** [comment]
It may actually slow things down as it can interfere with the optimizer. What we have observed is that lots of inline code can cause sudden changes in performance when even tiny changes are made in the code as it trips the optimizer into a cache miss. Drives people nuts trying to track it down as even a debug statement can cause the miss.

So we tend to stay away from inline except for very small, linear code snippets. I'd suggest doing so here as well unless you have something really short and without any conditionals in it.

**Comment by rhc54** [comment]
Testing these changes on https://github.com/open-mpi/ompi/pull/4623

**Comment by karasevb** [comment]
@rhc54 I will return to #434 later, and recheck the `inline` influence

**Comment by artpol84** [comment]
@rhc54 do you remember what compiler[s] had this issue?
My naive understanding is that in principle inline shouldn't bring any issues unless compiler has problems.

**Comment by rhc54** [comment]
Afraid it failed again. There is just something in the way the code is written that is causing it to fail when embedded in a deeper tree. I can take a crack at it tomorrow as well.

**Comment by rhc54** [comment]
@hjelmn @bwbarrett  can probably answer your question better than I as they were more involved in tracking down the wild performance swings. I don't think it is a compiler "bug" so much as an issue that inline code blocks optimization paths, thus causing cache lines to become brittle.

**Comment by hjelmn** [comment]
Not a compiler bug. Excessive inlining of common code paths can cause the instruction cache to thrash. I noticed this while optimizing vader. Deleting a couple instructions from the code path (should always speed things up, right?) caused the latency to spike. PAPI showed that the average icache miss went from 0.4 per MPI_Send to 10+. It is completely non-intuitive but unless you can show a very real improvement it is best to avoid excessive inlining.

**Comment by hjelmn** [comment]
FWIW this applies to macros as well.

**Comment by artpol84** [comment]
hmmm, 
* were you able to pinpoint a problem in the assembly code? 
* Does that mean that if you will copy-paste the content of the inline/macro into the place where they are invoked you will get the same issue? Otherwise it doesn't make sense to me.

**Comment by hjelmn** [comment]
@artpol84 Saw nothing wrong with the assembly. I had simply caused the common paths through ob1 and vader to cause the icache to thrash. I ended up removing a lot of inlining in vader to fix the issue. Yes, there are more instructions now but with less duplicated code the cache can stay hot.

**Comment by hjelmn** [comment]
@artpol84 And, yes. C&P of the code causes the same issue. I also bloats the binary code.

**Comment by artpol84** [comment]
So you mean that the following case was an issue:

```C
...
if( condition ){
    func();
}
<code below>
```
if `func()` is long and compiler assumes `taken` by default you will get only a function call in the iCache and the rest will be `<code below>`. But if you will copy/paste you'll get icache filled with those instructions.

It sound very case-specific. I mean if your fastpath is usually `taken` then I still see the benefit of C&P the code there. If it's 50/50 - then yes it will be a problem.

Am I getting it right?

**Comment by artpol84** [comment]
@hjelmn I read carefully what you wrote and I think that my previous question perhaps extra. Thanks.

**Comment by hjelmn** [comment]
@artpol84 Ok. Hope it makes sense. It really comes down to needing a high-level understanding of flow through your code. No compiler can do this unfortunately and the code flow can be complicated to work out so it is best to be cautious when inlining.

**Comment by rhc54** [comment]
Looks like the root cause of the issue here is that there is something not quite correct in the way the dstore_mod_t's are being handled. Simply removing one of them solves the problem.

**Comment by rhc54** [comment]
@karasevb @artpol84 After playing with this on the OMPI repo and talking about this situation on the telecon today, I would like to ask that we revert c4957fa in place of this PR. It appears that the move to function pointers has introduced a recursion loop that is breaking at least the optimizer in the PGI compiler, and may impact others (since we don't test every version of every compiler out there). I know there is an aesthetics preference in retaining that modification, but it isn't something that impacts actual code operation.

I think it would be fine if you chose to continue working with it on a branch, and I think we all are open to someday returning to it. However, it is holding up the release process on both PMIx and OMPI.

Would you please push a revert of that commit to the master branch? We can then go ahead with a PR of the remaining code to the v2.1 branch.

**Comment by artpol84** [comment]
@rhc54  I am not sure what do you mean "that we revert c4957fa in place of this PR".  Can you rephrase it and explain in more details what are you proposing?

We don't want to hold OMPI release, however we invested noticeable amount of time into this change. Before reverting a good piece of work I would like to see more detailed explanation of what we observed. My impression from https://github.com/open-mpi/ompi/pull/4606 was the you have more guess than a real detailed understanding.
Also I honestly can't figure out what has been changed in https://github.com/open-mpi/ompi/pull/4606 to make it pass the CI. PR is huge and I don't see what exactly helped.

Unfortunately I was unable to join the call today, however I'm open to have a separate call to discuss this particular issue.

**Comment by artpol84** [comment]
In other words: are we talking about cosmetic change to make compiler happy or you want to revert the whole work and release with broken dstore?

**Comment by artpol84** [comment]
Checking open-mpi/ompi#4606 I see that function pointers are there but CI is happy:
* https://github.com/open-mpi/ompi/pull/4606/commits/5d4c5621a6b4bc72047a6f67bf069452f8015085#diff-7b422726a1247e9038a793edc7fc8efcR92
* https://github.com/open-mpi/ompi/pull/4606/commits/5d4c5621a6b4bc72047a6f67bf069452f8015085#diff-887eefeed6fcbb126e3b59c91c01f616R135

What has been done to achieve this?

**Comment by rhc54** [comment]
I don't know what you are talking about - I reverted the function pointer code completely out of that PR. It is the PMIx master minus the commit we discussed elsewhere. This is why it now builds.
